lib/patientFinder/utils.rb

```diff
@@ -0,0 +1,165 @@
+        
+        #find matches for test patients
+        def self.get_MRNs(master_records, test_patients)
+    
+            #store the results
+            tests_to_MRNs= Hash.new
+            
+            #iterate through master records and find candidates 
+            test_patients.each { |test_patient|
+                #each test patient might have one or more MRN candidates
+                candidates = get_candidates_for_test(master_records, test_patient.record)
+                
+                #resolve cadidates to unique MRNs - with other words resolve the candidates
+                get_MRN_for_tests(candidates.map(&:medical_record_number), test_patient, tests_to_MRNs)
+            }
+    
+            tests_to_MRNs
+        end
+        
+        #resolves ambiguos candidates to a unique MRN 
+        def self.get_MRN_for_ambiguous(candidate_MRNs)
+            AMBIGUOUS_PATIENTS.each_pair {  |ambiguous_entries, mrn| 
+                return mrn if (candidate_MRNs-ambiguous_entries).empty?
+            }
+            "too ambiguous #{candidate_MRNs}"
+        end
+    
+        #resolves the candidates to the test files (one test file - one candidate - when multiple candidates are found manual anotation is used)
+        def self.get_MRN_for_tests(candidate_MRNs, test_record, test_to_MRN_hash)
+            
+            if(candidate_MRNs.length==1)
+                test_to_MRN_hash[test_record.fileName]=candidate_MRNs.first
+            else
+                test_to_MRN_hash[test_record.fileName]=get_MRN_for_ambiguous(candidate_MRNs)
+            end
+        end    
+    
+        #finds sections with data for a record (out of a array of given sections)    
+        # def self.get_non_empty_sections_for_record(sections, record)
+        #     sections_with_data=Array.new
+        #     sections.each{ |section|
+        #         if !(record[section].nil? || record[section].empty?)
+        #             sections_with_data<< section
+        #         end
+        #     }
+        #     sections_with_data
+        # end
+    
+        #a utility that iterates through a set of master records to find possible matches (candidates) for a test cord
+        def self.get_candidates_for_test(master_records,test_record)
+            
+            candidates=Array.new
+            # candidates
+            master_records.each { |master_record|
+            
+                #simple conditions that will reduce the record pool
+                if(master_record.gender==test_record.gender && master_record.ethnicity[:code]==test_record.ethnicity[:code] && master_record.race[:code] == test_record.race[:code])
```

Comment:
At minimum, I would prefer to have this broken up on multiple lines. We don't enforce an 80-character limit currently (partly because we inherited a lot of code that would violate it), but it's still a nice thing to target. So, e.g.

    if (master_record.gender == test_record.gender &&
        master_record.ethnicity[:code] == test_record.ethnicity[:code] &&
        master_record.race[:code] == test_record.race[:code])

At this indentation level the second line of that still hits 83 (which is one reason I usually prefer 2-space indentation to 4-space).

I would also consider a more significant refactor, possibly something like this:

    demo = Struct.new(:gender, :ethnicity, :race)
    get_demo = -> (obj) do
      demo.new(obj.gender, obj.ethnicity[:code], obj.race[:code])
    end
    if (get_demo.call(master_record) == get_demo.call(test_record))